File: XLua/XLuaObject.cpp
#include "Common.h"

// =====================================================================================
//
// Lua Manager (2)
// 
// =====================================================================================

XLuaObject::XLuaObject ()
	: state(NULL), errorQueue(), printQueue() {
}

XLuaObject::~XLuaObject () { 
	if (state) {
		state->UnbindRd(this);
		state = NULL;
	}
}

//-----------------------------------------------------------------------------

const std::string& XLuaObject::GetError () {
	if (!errorQueue.empty()) {
		errorString = errorQueue.front();
		errorQueue.pop();

		return errorString;
	}

	errorString.clear();
	return errorString;
}

const std::string& XLuaObject::GetPrint () {
	if (!printQueue.empty()) {
		printString = printQueue.front();
		printQueue.pop();

		return printString;
	}

	printString.clear();
	return printString;
}

void XLuaObject::RaiseError (const std::string& str) {
	if (errorQueue.size() >= 20) {
		errorQueue.pop();
	}

	errorQueue.push(str);
	lastErrorString = str;

	if (rdPtr == NULL) return;

	if (errMode == 0) {
		callRunTimeFunction(rdPtr, RFUNCTION_GENERATEEVENT, 0, 0);
		//rdPtr->rRd->GenerateEvent(0);
	}
	else if (errMode == 1) {
		callRunTimeFunction(rdPtr, RFUNCTION_PUSHEVENT, 0, 0);
		//rdPtr->rRd->PushEvent(0);
	}
}

void XLuaObject::RaisePrint (const std::string& str) {
	if (printQueue.size() >= 2000) {
		printQueue.pop();
	}
	printQueue.push(str);
	lastPrintString = str;

	if (rdPtr == NULL) return;

	if (printMode == 0) {
		callRunTimeFunction(rdPtr, RFUNCTION_GENERATEEVENT, 2, 0);
		//rdPtr->rRd->GenerateEvent(2);
	}
	else if (printMode == 1) {
		callRunTimeFunction(rdPtr, RFUNCTION_PUSHEVENT, 2, 0);
		//rdPtr->rRd->PushEvent(2);
	}
}

//-----------------------------------------------------------------------------

void XLuaObject::BindState (int sid) {
	//XLuaGlobal* glob = GetLG(rdPtr);
	
	if (state != NULL) {
		state->UnbindRd(this);
	}

	state = XLuaGlobal::Get().GetState(sid);
	if (state == NULL) {
		XLuaGlobal::Get().CreateState(sid);
		state = XLuaGlobal::Get().GetState(sid);
		stateId = sid;
	}

	if (state != NULL) {
		state->BindRd(this);
	}
}

void XLuaObject::UnbindState () {
	if (state != NULL) {
		state->UnbindRd(this);
		state = NULL;
	}
}

std::string XLuaObject::GetEmbeddedScript (const char* name) {
	IScripts si = scripts.find(name);
	if (si == scripts.end())
		return "";

	return si->second.script;
}

//-----------------------------------------------------------------------------

LPRO xlua_get_object (lua_State* L, int objid) {
#ifdef XLUA_LEGACY
	XLuaState* state = XLuaGlobal::Get().GetStateByState(L);
	
	XLuaState::IObjects iter = state->objects.find(objid);
	if (iter == state->objects.end()) {
		return NULL;
	}

	return iter->second.obj;
#else
	return 0;
#endif
}

LPRH xlua_get_run_header (lua_State* L) {
	XLuaState* state = XLuaGlobal::Get().GetStateByState(L);

	return state->rdList.front()->rdPtr->rHo.hoAdRunHeader;
}

File: XLua-Java/src/XLuaObject.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

import java.util.*;
import Objects.*;
import Extensions.*;

/**
 *
 * @author admin
 */
public class XLuaObject {

    public static final int COND_ERROR = 0;
    public static final int COND_PRINT = 2;

    public static final int MODE_DEFAULT = 0;
    public static final int MODE_IMMEDIATE = 1;
    public static final int MODE_QUEUED = 2;
    public static final int MODE_EXPRESSION = 3;
    
    public XLuaState state;

    // public LPRDATA rdPtr;
    public CExtension ho;

    public Queue<String> errorQueue;
    public Queue<String> printQueue;
    public String errorString = "";
    public String lastErrorString = "";
    public String printString = "";
    public String lastPrintString = "";

    public int errorMode;
    public int printMode;
    public boolean runScript;
    public boolean autoReg;
    public boolean useBacktrace;

    public int stateId;

    public Map<String, XLuaScriptRecord> scripts;

    public XLuaObject() {
        errorQueue = new PriorityQueue<String>();
        printQueue = new PriorityQueue<String>();
        scripts = new HashMap<String, XLuaScriptRecord>();
    }

    void raiseError (String str) {
        raiseError(str, MODE_DEFAULT);
    }

    void raiseError(String str, int mode) {
        System.out.println("Raise Error: " + str);
        if (errorQueue.size() >= 20) {
            errorQueue.poll();
        }

        errorQueue.offer(str);
        lastErrorString = str;

        if (ho == null) {
            return;
        }

        switch (mode) {
        case MODE_DEFAULT:
        case MODE_IMMEDIATE:
            ho.generateEvent(COND_ERROR, 0);
            break;
        case MODE_QUEUED:
        case MODE_EXPRESSION:
            ho.pushEvent(COND_ERROR, 0);
            break;
        }
    }

    void raisePrint(String str) {
        if (printQueue.size() >= 20) {
            printQueue.poll();
        }

        printQueue.offer(str);
        lastPrintString = str;

        if (ho == null) {
            return;
        }

        // TODO: Eventually handle errMode - thread safety issues
        ho.pushEvent(COND_PRINT, 0);
    }

    String getError() {
        if (!errorQueue.isEmpty()) {
            errorString = errorQueue.poll();
            return errorString;
        }

        errorString = "";
        return errorString;
    }

    String getPrint() {
        if (!printQueue.isEmpty()) {
            printString = printQueue.poll();
            return printString;
        }

        printString = "";
        return printString;
    }

    void bindState(int sid) {
        if (state != null) {
            state.UnbindRd(this);
        }

        XLuaGlobal glob = XLuaGlobal.Get();
        state = glob.GetState(sid);
        if (state == null) {
            glob.CreateState(sid);
            state = glob.GetState(sid);
            stateId = sid;
        }

        if (state != null) {
            state.BindRd(this);
        }
    }

    void unbindState() {
        if (state != null) {
            state.UnbindRd(this);
            state = null;
        }
    }

    String GetEmbeddedScript (String name) {
        if (scripts.containsKey(name)) {
            return scripts.get(name).script;
        }
        
        return "";
    }
}
